### PR TITLE
Reduce pointer indirection for global list_t variables

### PR DESCRIPTION
**Pull request by fangfufu**
Converted cmdline_conf, connection_list, outgoing_list from
pointer-to-structs to structs.

Created list_empty_list for these structs. This is necessary,
because list_delete_list frees the supplied list_t pointer.

Part of https://github.com/gsliepen/tinc/issues/294

### CODE REVIEW

**Comment by fangfufu** [comment]
The tests fail for some reason. It would be helpful if someone provides a bit of feedback on this.

**Comment by hg** [comment]
Try this patch.

```diff
diff --git a/src/conf.c b/src/conf.c
index 6b5f1093..90bd3697 100644
--- a/src/conf.c
+++ b/src/conf.c
@@ -44,7 +44,7 @@ list_t cmdline_conf = {
 	.head = NULL,
 	.tail = NULL,
 	.count = 0,
-	.delete = NULL,
+	.delete = (list_action_t)free_config,
 };
 
 static int config_compare(const config_t *a, const config_t *b) {
diff --git a/src/connection.c b/src/connection.c
index 20dc5de7..0c5e7ef0 100644
--- a/src/connection.c
+++ b/src/connection.c
@@ -35,7 +35,7 @@ list_t connection_list = {
 	.head = NULL,
 	.tail = NULL,
 	.count = 0,
-	.delete = NULL,
+	.delete = (list_action_t) free_connection,
 };
 
 connection_t *everyone;
```

**Comment by fangfufu** [comment]
Thanks, @hg

File: src/list.c [hunk 1]
```diff
@@ -184,11 +184,14 @@ void *list_get_tail(list_t *list) {
 
 /* Fast list deletion */
 
-void list_delete_list(list_t *list) {
+void list_empty_list(list_t *list) {
 	for(list_node_t *node = list->head, *next; next = node ? node->next : NULL, node; node = next) {
 		list_free_node(list, node);
 	}
+}
```
Comment (gsliepen) [inline src/list.c:191]: You should also set `list->head` and `list->tail` to `NULL` afterwards, and `list->count` to zero.

**Review by gsliepen — CHANGES_REQUESTED** [review]

File: src/list.c (hunk 1 above)
Comment (fangfufu) [reply src/list.c:195]: Done.

**Comment by fangfufu** [comment]
Resolving the conversation does not appear to get rid off the "change requested", @splitice.

**Comment by gsliepen** [comment]
Cherry-picked as 0871c3095151bce6a4031a2662aa51b7193b855c.

**Pull request closed without merging**